File: Python/kakao합승택시요금.py
import math
def solution(n, s, a, b, fares):
    INF = math.inf
    answer = math.inf
    arr = [[0 if i == j  else INF for i in range(n)] for j in range(n)]
    for i in range(len(fares)):
        x, y, w = fares[i]
        arr[x - 1][y - 1] = w
        arr[y - 1][x - 1] = w

    for i in range(n):
        for j in range(n):
            for k in range(n):
                arr[j][k] = min(arr[j][i] + arr[i][k], arr[j][k])

    for i in range(n):
        val = arr[s - 1][i] + arr[i][a - 1] + arr[i][b - 1]
        answer = min(answer, val)
    
    return answer

solution(7, 3, 4, 1, 	[[5, 7, 9], [4, 6, 4], [3, 6, 1], [3, 2, 3], [2, 1, 6]])

File: Java/KAKAO합승택시요금.java
public class KAKAO합승택시요금 {

	public int solution(int n, int s, int a, int b, int[][] fares) {
		final int INF = 1000001;
        int answer = Integer.MAX_VALUE;
        
        int[][] arr = new int[n][n];
        for(int i = 0; i < n; i++) {
        	for(int j = 0; j < n; j++) {
        		if(i != j) {
        			arr[i][j] = INF;
        		}
        	}
        }
        for(int i = 0; i < fares.length; i++) { // 양방향이기 때문에 모두 표시
        	arr[fares[i][0] - 1][fares[i][1] - 1] = fares[i][2];
        	arr[fares[i][1] - 1][fares[i][0] - 1] = fares[i][2];
        }
        
        for(int i = 0; i < n; i++) { // 플로이드-와샬 알고리즘 사용
        	for(int j = 0; j < n; j++) {
        		for(int k = 0; k < n; k++) {
        			arr[j][k] = Math.min(arr[j][k], arr[j][i] + arr[i][k]);
        		}
        	}
        }
        for(int i = 0; i < n; i++) { // 모든 지역 검사
    		int val = arr[s - 1][i] + arr[i][a - 1] + arr[i][b - 1];
    		answer = Math.min(answer, val);
        }
        return answer;
    }
}
